pkg/internal/vault/vault.go

```diff
@@ -355,3 +340,29 @@ func (v *Vault) requestTokenWithKubernetesAuth(client Client, kubernetesAuth *v1
 func (v *Vault) Sys() *vault.Sys {
 	return v.client.Sys()
 }
+
+func extractCertificatesFromVaultCertificateSecret(secret *certutil.Secret) ([]byte, []byte, error) {
+	parsedBundle, err := certutil.ParsePKIMap(secret.Data)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to decode response returned by vault: %s", err)
+	}
+
+	bundle, err := parsedBundle.ToCertBundle()
+	if err != nil {
+		return nil, nil, fmt.Errorf("unable to convert certificate bundle to PEM bundle: %s", err.Error())
+	}
+
+	var caPem []byte
+	if len(bundle.CAChain) > 0 {
+		caPem = []byte(bundle.CAChain[len(bundle.CAChain)-1])
+	} else {
+		caPem = []byte(bundle.IssuingCA)
+	}
+
+	crtPems := []string{bundle.Certificate}
+	if len(bundle.CAChain) > 1 {
+		crtPems = append(crtPems, bundle.CAChain[0:len(bundle.CAChain)-1]...)
+	}
```

Comment:
The assumption being made here/the change in behaviour isn't quite right.

The equivalent code it is replacing calls `bundle.ToPEMBundle()` to set `crtPems` instead.

Now it has been changed to set `crtPems` to `bundle.Certificate` (*just* the leaf), and to append `bundle.CAChain` *only* if it contains more than one certificate.

This is effectively assuming that `CAChain` will *always* contain a root CA (and also assuming that if only one certificate is present in `CAChain`, that it is a root certificate. In cases where users have configured their Vault backend to only be aware of the intermediate certificate (which is a fairly common way to configure your backend), this means the `tls.crt` entry will now no longer contain the intermediate certificate (thus breaking the chain up to any root certificates they may have in their trust store).

This logic needs adapting to actually inspect the certificates and determine whether they are root CAs to determine whether they should be used as the caPem or the crtPems.

I think this may also require some changes to the lines above as well:

```
	var caPem []byte
 	if len(bundle.CAChain) > 0 {
 		caPem = []byte(bundle.CAChain[len(bundle.CAChain)-1])
 	} else {
 		caPem = []byte(bundle.IssuingCA)
 	}
```

As I'm not too sure on the relationship between `CAChain` and `IssuingCA` 🤔

Reply:
~~The fix is being worked on in https://github.com/jetstack/cert-manager/pull/3978~~

The PR has been merged into four PRs:
- https://github.com/jetstack/cert-manager/pull/3982
- https://github.com/jetstack/cert-manager/pull/3983
- https://github.com/jetstack/cert-manager/pull/3984
- https://github.com/jetstack/cert-manager/pull/3985